rl_power_markets/benchmarks/markets/full.py

```diff
@@ -1,3 +1,220 @@
+        for i in generators:
+            for b in range(C):
+                lhs_gen += g[i, h, b]
+        for c in range(C):
+            lhs_dem += d[h, c]
+        m.addConstr(lhs_gen == lhs_dem, name=f"balance_{h}")
+    for i in generators:
+        for h in range(H):
+            m.addConstr(gp.quicksum(g[i, h, b] for b in range(C)) <= bigM[i] * u[i, h], name=f"cap_{i}_{h}")
+    for i in generators:
+        for h in range(1, H):
+            m.addConstr(u[i, h] - u[i, h-1] <= su[i, h], name=f"start_{i}_{h}")
+            m.addConstr(u[i, h-1] - u[i, h] <= sd[i, h], name=f"shut_{i}_{h}")
+            m.addConstr(su[i, h] + sd[i, h] <= 1, name=f"switch_{i}_{h}")
+    for i in generators:
+        for h in range(1, H):
+            m.addConstr(gp.quicksum(g[i, h, b] for b in range(C)) - gp.quicksum(g[i, h-1, b] for b in range(C)) <= generators[i]["RU"], name=f"rampup_{i}_{h}")
+            m.addConstr(gp.quicksum(g[i, h-1, b] for b in range(C)) - gp.quicksum(g[i, h, b] for b in range(C)) <= generators[i]["RD"], name=f"rampdown_{i}_{h}")
+    for i in generators:
+        UT = generators[i]["UT"]
+        DT = generators[i]["DT"]
+        for h in range(H - UT + 1):
+            m.addConstr(gp.quicksum(1 - u[i, t] for t in range(h, h+UT)) <= UT - 1, name=f"minup_{i}_{h}")
+        for h in range(H - DT + 1):
+            m.addConstr(gp.quicksum(u[i, t] for t in range(h, h+DT)) <= 1, name=f"mindown_{i}_{h}")
+    m.optimize()
+    return m, u, g, d
+
+class FullMarket:
+    def __init__(self, batch_size: int = 32) -> None:
+        self.batch_size = batch_size
+        self.num_hours = 24
+        
+        # Initialize generators data
+        self.generators = {
+            1: {"b":5.5, "CNL":80000, "g_max":6.584, "RU":1.317, "RD":1.317, "CSU":400000, "CSD":80000, "UT":24, "DT":24},
+            2: {"b":30.0, "CNL":6500, "g_max":5.760, "RU":1.152, "RD":1.152, "CSU":32500, "CSD":2850, "UT":20, "DT":20},
+            3: {"b":35.0, "CNL":2851, "g_max":3.781, "RU":1.512, "RD":1.512, "CSU":14250, "CSD":1850, "UT":16, "DT":16},
+            4: {"b":60.0, "CNL":1440, "g_max":3.335, "RU":1.334, "RD":1.334, "CSU":7200, "CSD":1440, "UT":10, "DT":10},
+            5: {"b":80.0, "CNL":1100, "g_max":3.252, "RU":1.951, "RD":1.951, "CSU":5500, "CSD":1200, "UT":8, "DT":8},
+            6: {"b":95.0, "CNL":620, "g_max":2.880, "RU":1.728, "RD":1.728, "CSU":3100, "CSD":1000, "UT":5, "DT":5},
+            7: {"b":100.0, "CNL":560, "g_max":2.748, "RU":2.198, "RD":2.198, "CSU":1120, "CSD":840, "UT":4, "DT":4}
+        }
+        
+        # Setup block structure for each generator
+        self.C = 5  # number of blocks
+        for i in self.generators:
+            step = (self.generators[i]["g_max"] - 0)/self.C
+            blocks = []
+            for b in range(self.C):
+                low = b * step
+                high = (b+1) * step
+                mid = 0.5*(low+high)
+                slope = self.generators[i]["b"] + 2 * 0.0001 * mid
```

Comment:
can we move consts that are not table to top of file? common bug vector with hard to notice effects